tiledb/sm/filesystem/s3/STSCredentialsProvider.cc

```diff
@@ -0,0 +1,233 @@
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines a vendored copy of the
+ * Aws::Auth::STSAssumeRoleWebIdentityCredentialsProvider class, updated to
+ * support getting a client configuration object.
+ *
+ * Changes made should be contributed upstream to the AWS SDK for C++ and when
+ * that happens, the vendored copy should be removed.
+ */
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifdef HAVE_S3
+
+#include "tiledb/sm/filesystem/s3/STSCredentialsProvider.h"
+#include "tiledb/common/common.h"
+
+#include <aws/core/client/SpecifiedRetryableErrorsRetryStrategy.h>
+#include <aws/core/config/AWSProfileConfigLoader.h>
+#include <aws/core/platform/Environment.h>
+#include <aws/core/platform/FileSystem.h>
+#include <aws/core/utils/FileSystemUtils.h>
+#include <aws/core/utils/StringUtils.h>
+#include <aws/core/utils/UUID.h>
+#include <aws/core/utils/logging/LogMacros.h>
+#include <string.h>
+#include <climits>
+#include <cstdlib>
+#include <fstream>
+
+using namespace Aws::Utils;
+using namespace Aws::Utils::Logging;
+using namespace Aws::Auth;
+using namespace Aws::Internal;
+using namespace Aws::FileSystem;
+using namespace Aws::Client;
+using Aws::Utils::Threading::ReaderLockGuard;
+using Aws::Utils::Threading::WriterLockGuard;
+
+static const char STS_ASSUME_ROLE_WEB_IDENTITY_LOG_TAG[] =
+    "STSAssumeRoleWithWebIdentityCredentialsProvider";
+static const int STS_CREDENTIAL_PROVIDER_EXPIRATION_GRACE_PERIOD = 5 * 1000;
+
+namespace tiledb::sm::filesystem::s3 {
+STSAssumeRoleWebIdentityCredentialsProvider::
+    STSAssumeRoleWebIdentityCredentialsProvider(
+        Aws::Client::ClientConfiguration config)
+    : m_initialized(false) {
+  // check environment variables
+  Aws::String tmpRegion = Aws::Environment::GetEnv("AWS_DEFAULT_REGION");
```

Comment:
As a follow up, can we improve the region support here? It's be nice to allow using the user set config of `vfs.s3.region`. It's a little unexpected for users using any of the credentials providers that the region for sta is only read from the env or profile.

We encountered this oddity with a customer running in AWS China and struggling to understand why STS was trying to hit `us-east-1` instead of the `cn-notth-1` that was needed and configured via the TileDB config for region.

Reply:
Fixed. If the config has a region specified, it will not be overridden. I will make sure to include this to my upstream fix.

Reply:
> If the config has a region specified, it will not be overridden

After further thought this won't work because the region of a config always gets initialized with a value. I think the best thing to do here is always get the region in the client provider from the config, and in a subsequent PR update the default value of `vfs.s3.region` from `us-east-1` to letting the AWS SDK choose, which would align the SDK's and TileDB's view of the default AWS region.

Reply:
Updated to always get the region from the client config. Will later open a story to consider changing the default value of `vfs.s3.region`.

Reply:
> In thinking on this, won't this change now mean that if a user set their profile or env variables they are ignored?

Yes, but this change in the credentials provider is more likely to be accepted upstream. This in combination with changing the default value of `vfs.s3.region` (SC-55556) will achieve the sequence you described.